chapters/image_accceses.md

```diff
@@ -0,0 +1,97 @@
+
+```c++
+struct ImageType {
+  bool depth;
+  bool arrayed;
+  bool sampled;
+  bool multisampled;
+  Format format;
+  int width;
+  // other metadata ...
+  void* texels; // data of the image (which is mutable)
+}
+```
+
+The above `OpVariable` is read-only because we are not allowed to adjust its format, width, or multisampling state ever.
+
+# Sampled Images
+
+When using a sampler, you must have a `OpTypeSampledImage` object, this will be directly pointing to a `OpTypeImage` object.
+
+To access the sampled image you will use a `OpImage*Sample*` instruction (ex. `OpImageSampleImplicitLod`) to get a `OpSampledImage`. (It can also use a `OpImage*Gather` as well)
+
+![image_access_sampled_image.png](../images/image_access_sampled_image.png)
+
+## Fetch
+
+There are also `OpImageFetch` (and `OpImageSparseFetch`) instructions which work similar as `OpImage*Sample*`.
+
+While the `OpImage*Sample*` takes float coordinates and does a lookup to get the texel, the `OpImageFetch` takes direct image coordinates to the exact texel to return.
+
+This means an `OpImageFetch` will directly access through an `OpImage` and doesn't need an `OpSampledImage` object.
+
+# Non-sampled image
+
+When not using a sampler (ex. storage image), you can directly use an `OpTypeImage` object.
+
+> Note you will **not** be using an `OpImage` when accessing an image without a sampler. The `OpImage` is designed to point to the image of the sampler/image combination.
+
+The `OpImageRead` and `OpImageWrite` will directly go to the `OpLoad` into the variable.
+
+![image_access_storage_image.png](../images/image_access_storage_image.png)
+
+> The `OpImageWrite` calls `OpLoad` but this is **not** an image "read access". The `OpLoad` just grabs the reference, but there is only a "write access" here on the image.
+
+# Image Queries
+
+There are some `OpImageQuery*` instructions that are designed to directly access an `OpTypeImage` object. The image itself is not accessed.
+
+## OpImageTexelPointer
+
+Before you do an operation such as an atomic image load, you need to get a pointer for the memory.
+
+The `OpImageTexelPointer` allows you take a texel inside an image, and create a pointer to it which can be accessed by an atomic image operation.
+
+Example: `OpAtomicLoad -> OpImageTexelPointer -> OpVariable`
+
+To see this in practice we can view the difference between a `imageAtomicStore` and `imageStore` call
+
+```swift
+%1 = OpImageTexelPointer %OpTypePointer %atomic_image %coor %sample
+     OpAtomicStore %1 %_ %_ %scalar_value
+
+%2 = OpLoad %OpTypeImage %image
+     OpImageWrite %2 %coor %texel
+```
+
+While the `OpLoad` returns a `OpTypeImage` that we store the new texel, with `OpImageTexelPointer` we get a direct pointer to the texel data and can store scalar value like `OpConstant %int 0`. This is posssible because the `Coordinate` operand is supplied in the `OpImageTexelPointer` instead aftewards in the actual "write" operation.
```

Comment:
```suggestion
While the `OpLoad` returns a `OpTypeImage` that we store the new texel, with `OpImageTexelPointer` we get a direct pointer to the texel data and can store scalar value like `OpConstant %int 0`. This is posssible because the `Coordinate` operand is supplied in the `OpImageTexelPointer` instead afterwards in the actual "write" operation.
```

Reply:
Might be worth saying explicitly (I think this is true?): you're not expected to do any pointer math on the pointer returned by `OpImageTexelPointer`.  To get a pointer to a different texel, you should call `OpImageTexelPointer` again, with a different coordinate.

Reply:
aww, never thought about that, added!